src/notifications/signal.h

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (C) 2021 by the Widelands Development Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+ *
+ */
+
+#ifndef WL_NOTIFICATIONS_SIGNAL_H
+#define WL_NOTIFICATIONS_SIGNAL_H
+
+#include <cassert>
+#include <functional>
+#include <list>
+#include <memory>
+#include <set>
+
+#include "base/macros.h"
+#include "base/wexception.h"
+
+namespace Notifications {
+
+/**
+ * Whether a new subscriber will be the last (default)
+ * or first to be invoked when the signal is triggered.
+ */
+enum class SubscriberPosition { kFront, kBack };
+
+/** Class that allows attaching callback functions to an object. */
+template <typename... Args> class Signal {
+public:
+	/** Wrapper around a callback function. */
+	struct SignalSubscriber {
+		explicit SignalSubscriber(const Signal& p, const std::function<void(Args...)>& c)
+		   : parent_(p), callback_(c) {
+		}
+		~SignalSubscriber() {
+			parent_.unsubscribe(this);
+		}
+
+		const Signal& parent_;
+		const std::function<void(Args...)> callback_;
+
+		DISALLOW_COPY_AND_ASSIGN(SignalSubscriber);
+	};
+
+	/** Invoke all the signal's subscribers' callback functions. */
+	void operator()(Args... args) const {
+		for (const auto& s : all_subscribers_) {
+			s->callback_(args...);
+		}
+	}
+
+	/** Create a subscriber with a user-defined lifetime. */
+	std::unique_ptr<SignalSubscriber>
+	subscribe(const std::function<void(Args...)>& callback,
```

Comment:
pass std::function by value, it is usually a better choice: https://stackoverflow.com/questions/18365532/should-i-pass-an-stdfunction-by-const-reference#:~:text=If%20there%20is%20any%20possibility,as%20efficient%20as%20a%20const%26%20.